Performance Refactor on GenerateTransactionID SQL (3108)

sfount opened the pull request:
This PR represents the initial implementation of the performance improvement proposal https://github.com/IMA-WorldHealth/bhima-2.X/issues/3108 as discussed in https://github.com/IMA-WorldHealth/bhima-2.X/issues/3002. 

The referenced issues do a good job of explaining the details and analysis, in brief this PR: 
* Updates the GenerateTransactionID SQL function for improved performance 
* Implements an upgrade script to consider current production systems 
* Updates server controllers responsible for populating the additional data column required

I am not 100% on the latest SQL migration standards, what is included is trying to replicate the standard I have seen. 

**Please note: This pull request changes a fundamental method that is relied on by every financial entity in the system, it should be thoroughly vetted before integrating with the production code base.**

sfount committed:
perf(transactions): migrate SQL, optimised ID gen

This commit includes database breaking code and bumps the database
version to v0.9.0. It introduces SQL with an optimised core
`GenerateTransactionId` function, this function uses an indexed
MEDIUMINT field to produce an ID in O(1) vs. the O(N) complexity
of the legacy function.

This should have significant performance implications on large datasets.

sfount committed:
perf(transactions): write perf. column to journals

This commit updates everything that accesses the `posting_journal` and
`general_ledger` to include the new `trans_id_reference_number` column.

It also updates `schema.sql` with the updated v0.9.0 definition.

sfount committed:
chore(transactions): clean up ID SQL; minor fixes

This commit updates test data files to ensure the mock data mirrors how
the application would actually function with the new Transaction ID
generation. It also cleans up recently introduced SQL.

jniles commented on server/models/functions.sql:
```diff
@@ -66,28 +66,43 @@ END $$
   DECLARE trans_id_length TINYINT(1) DEFAULT 4;
   RETURN (
-    SELECT CONCAT(abbr, IFNULL(MAX(increment), 1)) AS id
+    SELECT CONCAT(
```
Great rewrite - this is clear than the previous version as well!

I think you can remove `trans_id_length`, since it has no more bearing on this function.  Note that this will need to be done in the migration script as well for completeness.

jniles commented on client/src/modules/journal/journal.service.js:
```diff
@@ -74,6 +74,8 @@ function JournalService(Api, AppCache, Filters, Periods, Modal, bhConstants, Tra
       .then(service.util.unwrapHttpResponse);
   }
 
+  // @TODO(sfount) new rows will need to access the transaction shared attribute
```
Why will new rows need access to this attribute?

jniles reviewed (commented):
Looks good to me!  One quick thing to clean up and then I'm satisfied with it.

A future optimization we might do is to remove the INDEX on trans_id, and instead parse the incoming transaction id, then match on trans_id_reference_number, then match on the result of that.  It would probably be faster since it would be looking at an integer and save use a heap of space by avoiding the trans_id index.

jniles committed:
chore: remove unused variable declarations

This commit cleans up the GetTransactionId() function by removing unused
variable declarations.

jniles committed:
Merge remote-tracking branch 'upstream/master' into perf-transaction-id-generation-3108

jniles committed:
fix: test dataset bugs

This commit fixes a few bugs in the test dataset, whereby a cash payment
had both is_caution set and an invoice linked.  It also added an
additional line to the invoice table to allow for better testing.
